### PR TITLE
gh-109870: Dataclasses: batch up exec calls

### LINKED ISSUE #109870: Combine dataclass exec calls
# Feature or enhancement

### Proposal:

[As mentioned here](https://discuss.python.org/t/improving-dataclasses-startup-performance/15442/18), dataclass definition can possibly be sped up by combining the multiple `exec` calls into one.

I have two versions, and I found that this results in either a 1.1-1.22 ~~1.26, or 1.37 speedup versus main~~ (old time / new time)

~~[version 1](https://github.com/python/cpython/commit/e7c42bd21cf0f12ebfd88744dfde70b78877b1ac) combines the exec calls by passing a list of name/txt/locals from each function definition.
[version 2](https://github.com/python/cpython/commit/10a72b7621b4a6580db1b29d15f259109a701c08) simplifies the final exec call to avoid `__create_fn__`, but I'm not experienced to know if that is safe.~~ (see below for open pr)

@ericvsmith Let me know if this is a suitable change, and if so, which version we should open a pr based on.

Thank you!

### Has this already been discussed elsewhere?

No response given

### Links to previous discussion of this feature:

https://discuss.python.org/t/improving-dataclasses-startup-performance/15442/18

### Linked PRs
* gh-110186
* gh-110851

### PR DESCRIPTION
**Pull request by ericvsmith**
Fixes #109870

Only call exec once per dataclass, instead of once per added method.

* Issue: gh-109870

### CODE REVIEW

File: Lib/dataclasses.py [hunk 1]
```diff
@@ -446,32 +446,94 @@ def _tuple_str(obj_name, fields):
...
+        # class?  The default is to not overwrite a function that already
+        # exists.
+        if unconditional_add:
+            self.unconditional_adds[name] = True
+
+        self.names.append(name)
+
+        if return_type is not MISSING:
+            self.locals['__dataclass_return_type__'] = return_type
+            return_annotation = '->__dataclass_return_type__'
```
Comment (ssweber) [inline Lib/dataclasses.py]: Maybe add a comment to make jt clear that only `__init__` uses this? If another function is generated (in hypothetical future) with a return annotation, then the combined locals would overwrite init's `None`, correct?

Comment (ssweber) [reply Lib/dataclasses.py]: To mitigate this, my pr had:

```
        fn_name = name.replace("__", "")
        locals[f'__dataclass_{fn_name}_return_type__'] = return_type
        return_annotation = f'->__dataclass_{fn_name}_return_type__'

**Comment by ericvsmith** [comment]
One thing I don't understand is whey the tests for pprinting some dataclass reprs changed. I've double-checked and don't see anything that would have affected that.

**Comment by ssweber** [comment]
> One thing I don't understand is whey the tests for pprinting some dataclass reprs changed. I've double-checked and don't see anything that would have affected that.

Things I've tried, that make no difference:

1. Copied old __repr__ code def verbatim
2. Explicitly call _recursive_repr(fn) on created object, rather than use decorator

Will keep chipping away at it.

**Comment by ssweber** [comment]
```
import pprint
import dataclasses as dataclasses

@dataclasses.dataclass
class dataclass1:
    field1: str
    field2: int
    field3: bool = False
    field4: int = dataclasses.field(default=1, repr=False)

dc = dataclass1("some fairly long text", int(1e10), True)
formatted = pprint.pformat(dc, width=60, indent=4)
pprint.pprint(formatted)
```

3.12 prints:
```
("dataclass1(field1='some fairly long text',\n"
 '           field2=10000000000,\n'
 '           field3=True)')
```

this pr prints:

```
"dataclass1(field1='some fairly long text', field2=10000000000, field3=True)"
```

notice the lack of parenthesis. Trying to figure out why that is.

**Comment by ssweber** [comment]
ah, it's line 193 in pprint.py. `__create_fn__` won't be in 

```
__create_fn__" in object.__repr__.__wrapped__.__qualname__)
```

a part of the larger check:

```
            elif (_dataclasses.is_dataclass(object) and
                  not isinstance(object, type) and
                  object.__dataclass_params__.repr and
                  # Check dataclass has generated repr method.
                  hasattr(object.__repr__, "__wrapped__") and
                  "__create_fn__" in object.__repr__.__wrapped__.__qualname__):
```

**Comment by ssweber** [comment]
`__create_fn__` got renamed to `create_fn`

File: Lib/dataclasses.py [hunk 2]
```diff
@@ -446,32 +446,94 @@ def _tuple_str(obj_name, fields):
...
+        #   self.x = x
+        #   self.y = y
+        #  def __repr__(self):
+        #   return f"cls(x={self.x!r},y={self.y!r})"
+        # return __init__,__repr__
+
+        txt = f"def create_fn({local_vars}):\n{fns_src}\n return {return_names}"
+        ns = {}
+        exec(txt, self.globals, ns)
+        fns = ns['create_fn'](**self.locals)
```
Comment (ssweber) [inline Lib/dataclasses.py]: Either keep `create_fn` and change pprint, or rename back to `__create_fn__`

That was a head-scratcher!

**Comment by ericvsmith** [comment]
@ssweber Thanks for the debugging! I'll do something to fix it. Probably rename back to `__create_fn__`, since that seems easiest.

File: Lib/dataclasses.py (hunk 2 above)
Comment (ericvsmith) [reply Lib/dataclasses.py]: Fixed.

File: Lib/dataclasses.py (hunk 1 above)
Comment (ericvsmith) [reply Lib/dataclasses.py]: Fixed.

**Comment by ssweber** [comment]
@ericvsmith, what’s your thought process in when/if to merge this? Im genuinely just interested.

**Comment by ericvsmith** [comment]
I want to clean it up a bit more. For example, the generated functions are specified with no indents (or maybe on level of indentation, don't recall), then another level is added later. It would be best just to generate the correct level of indentation in the first place. I doubt this change will result in any measurable difference, but I'd still like to do it. I think it will clean up some of the logic, at least.

**Pull request merged**